geode-core/src/distributedTest/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceConstraintsDistributedTest.java

```diff
@@ -0,0 +1,595 @@
+  @Test
+  @Parameters({"LOCAL", "REPLICATE", "PARTITION"})
+  @TestCaseName("[{index}] {method}(RegionType:{0})")
+  public void queriesInFlightExecutedByServersShouldNotBeAffectedWhenMethodAuthorizerIsChanged(
+      RegionShortcut regionShortcut) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    String queryString = "<TRACE> SELECT object.getName() FROM /" + regionName + " object";
+
+    server.invoke(() -> {
+      // Set test query observer.
+      assertThat(ClusterStartupRule.getCache()).isNotNull();
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      TestQueryObserver queryObserver =
+          new TestQueryObserver(RestrictedMethodAuthorizer.class.getName(), Collections.emptySet());
+      QueryObserverHolder.setInstance(queryObserver);
+
+      // Execute query.
+      Query query = internalCache.getQueryService().newQuery(queryString);
+      @SuppressWarnings("unchecked")
+      SelectResults<QueryObject> result = (SelectResults<QueryObject>) query.execute();
+      assertThat(queryObserver.invocations.get()).isEqualTo(1);
+      assertThat(result.size()).isEqualTo(ENTRIES);
+
+      // Execute query again, it should fail as the restricted authorizer has been installed.
+      QueryObserverHolder.reset();
+      Query newQuery = internalCache.getQueryService().newQuery(queryString);
+      assertThatThrownBy(newQuery::execute)
+          .isInstanceOf(NotAuthorizedException.class)
+          .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "getName");
+    });
+  }
+
+  /**
+   * The test creates an index with a method invocation as part of the expression, changes
+   * the {@link MethodInvocationAuthorizer} to a custom one that still allows the methods part of
+   * the index expression and executes a region operation from the client that would cause an
+   * index mapping removal/addition.
+   * The operation should succeed, the index should still be valid and no errors should be logged.
+   */
+  @Test
+  @Parameters(method = "getRegionTypeAndOperation")
+  @TestCaseName("[{index}] {method}(RegionType:{0};Operation:{1})")
+  public void indexesShouldNotBeAffectedByMethodAuthorizerChangeAfterRegionOperationOnClientWhenIndexedExpressionContainsMethodsAllowedByTheNewAuthorizer(
+      RegionShortcut regionShortcut, Operation operation) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    QueryObject testValue = new QueryObject(999, "name_999");
+
+    server.invoke(() -> {
+      assertThat(ClusterStartupRule.getCache()).isNotNull();
+      InternalCache internalCache = ClusterStartupRule.getCache();
+
+      // Index is valid.
+      QueryService queryService = internalCache.getQueryService();
+      queryService.createIndex("NameIndex", "e.getName()", "/" + regionName + " e");
+      Index index = queryService.getIndex(internalCache.getRegion(regionName), "NameIndex");
```

Comment:
Here and in other tests, could we extract the index name "NameIndex" (or "IdIndex" in tests where that's used) to a variable, just to avoid having it hardcoded in two different places?

Reply:
Done!.